### PR TITLE
Cart Depot

### PR DESCRIPTION
**Pull request by RuoanJi**
Reopen a PR for this feature. There is another PR opened by @Whiteknight https://github.com/Keriew/augustus/pull/593 but I have made some updates since.

Details of the proposed feature can be found in the original PR https://github.com/Keriew/augustus/pull/414

### CODE REVIEW

File: src/window/building/depot.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#ifndef WINDOW_BUILDING_DEPOT_H
+#define WINDOW_BUILDING_DEPOT_H
+
+#include "common.h"
+#include "input/mouse.h"
+
+void window_building_depot_init();
```
Comment (crudelios) [inline src/window/building/depot.h:7]:
```suggestion
void window_building_depot_init(void);
```

File: src/window/building/depot.c [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+#include "window/building_info.h"
+
+static void order_set_source(int index, int param2);
+static void order_set_destination(int index, int param2);
+static void order_set_resource(int index, int param2);
+static void order_set_condition_type(int index, int param2);
+static void order_set_condition_threshold(int index, int param2);
+static void set_order_source(int depot_building_id, int building_id);
+static void set_order_destination(int depot_building_id, int building_id);
+static void set_order_resource(int depot_building_id, int resource_id);
```
Comment (crudelios) [inline src/window/building/depot.c:24]: The three last functions don't need a declaration. See above.

File: src/window/building/depot.c [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+#define DEPOT_BUTTONS_Y_OFFSET 204
+static generic_button depot_order_buttons[] = {
+    {100, 0, 26, 26, order_set_resource, button_none, 1, 0},
+    {100, 56, 284, 22, order_set_source, button_none, 2, 0},
+    {100, 82, 284, 22, order_set_destination, button_none, 3, 0},
+    {100, 30, 284, 22, order_set_condition_type, button_none, 4, 0},
+    {384, 30, 32, 22, order_set_condition_threshold, button_none, 5, 0},
+};
+
+static int storage_buildings_count() {
```
Comment (crudelios) [inline src/window/building/depot.c:51]: On functions, the `{` is on a different line. Also, don't forget the `void` on the paramenter list.

Comment (crudelios) [inline src/window/building/depot.c:51]: Also, you don't need to count all types. You can use `building_first_of_type(type)` and then increase `count` while there is a `b->next_of_type`.

File: src/window/building/depot.c [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+            count++;
+        }
+    }
+    return count;
+}
+
+#define ROW_HEIGHT 22
+#define MAX_VISIBLE_ROWS 15
+static void on_scroll(void);
+static scrollbar_type scrollbar = { 0, 0, ROW_HEIGHT * MAX_VISIBLE_ROWS, 432, MAX_VISIBLE_ROWS, on_scroll, 0, 4 };
```
Comment (crudelios) [inline src/window/building/depot.c:65]: Formatting: These four lines should be before any code.

File: src/window/building/depot.c [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+    {0, 0, 0, ROW_HEIGHT, button_none, button_none, 0, 0},
+};
```
Comment (crudelios) [inline src/window/building/depot.c:194]: This should also go to the top.

File: src/window/building/depot.c [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+    int index = 0, offset = scrollbar.scroll_position;
+    for (int i = 1; i < building_count(); i++) {
+        building *b = building_get(i);
+        if (b->type == BUILDING_GRANARY || b->type == BUILDING_WAREHOUSE) {
+            if (index == MAX_VISIBLE_ROWS) {
+                break;
+            }
+            if (offset > 0) {
+                offset--;
+            }
```
Comment (crudelios) [inline src/window/building/depot.c:212]: Formatting: `else` on the same line.

File: src/window/building/depot.c [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+        depot_select_storage_buttons[i].x = 0;
+        depot_select_storage_buttons[i].y = 22 * i;
+        depot_select_storage_buttons[i].width = 16 * (c->width_blocks - 2) - 4 -
+            (scrollbar.max_scroll_position > 0 ? 39 : 0);
+        depot_select_storage_buttons[i].height = ROW_HEIGHT;
+        depot_select_storage_buttons[i].left_click_handler = button_none;
+        depot_select_storage_buttons[i].parameter1 = 0;
+        depot_select_storage_buttons[i].parameter2 = 0;
+    }
+    for (int i = 1; i < building_count(); i++) {
```
Comment (crudelios) [inline src/window/building/depot.c:250]: You don't need to go through all buildings here either. See above.

File: src/window/building/depot.c [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+        building *b = building_get(i);
+        if (b->type == BUILDING_GRANARY || b->type == BUILDING_WAREHOUSE) {
+            if (index == MAX_VISIBLE_ROWS) {
+                break;
+            }
+            if (offset > 0) {
+                offset--;
+            }
```
Comment (crudelios) [inline src/window/building/depot.c:258]: Formatting: `else` on same line.

File: src/window/building/depot.c [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+    {18, 78, 214, 26, set_order_resource, button_none, 0, 7},
+    {232, 78, 214, 26, set_order_resource, button_none, 0, 8},
+    {18, 104, 214, 26, set_order_resource, button_none, 0, 9},
+    {232, 104, 214, 26, set_order_resource, button_none, 0, 10},
+    {18, 130, 214, 26, set_order_resource, button_none, 0, 11},
+    {232, 130, 214, 26, set_order_resource, button_none, 0, 12},
+    {18, 156, 214, 26, set_order_resource, button_none, 0, 13},
+    {232, 156, 214, 26, set_order_resource, button_none, 0, 14},
+    {18, 182, 214, 26, set_order_resource, button_none, 0, 15},
+};
```
Comment (crudelios) [inline src/window/building/depot.c:327]: To the top this goes!

File: src/window/building/figures.c [hunk 10]
```diff
@@ -362,6 +372,47 @@ static void draw_cartpusher(building_info_context *c, figure *f)
...
+    if (f->loads_sold_or_carrying > 0 && f->resource_id != RESOURCE_NONE) {
+        image_draw(image_group(GROUP_RESOURCE_ICONS) +
+            f->resource_id + resource_image_offset(f->resource_id, RESOURCE_IMAGE_ICON),
+            c->x_offset + 92, c->y_offset + 135, COLOR_MASK_NONE, SCALE_NONE);
+        text_draw_number(f->loads_sold_or_carrying, 'x', "", c->x_offset + 118, c->y_offset + 139, FONT_NORMAL_BROWN, COLOR_BLACK);
+    }
+
+    int resource_type = f->current_order.resource_type;
+    building* src = building_get(f->current_order.src_storage_id);
+    building* dst = building_get(f->current_order.dst_storage_id);
```
Comment (crudelios) [inline src/window/building/figures.c:394]: Formatting (previous two lines) `building *src/dst`.

**Review by crudelios — CHANGES_REQUESTED** [review]
Some things to fix.

Most of them are small formatting issues (I probably didn't catch them all though), but there is one or two more substancial things that I think need changing.

One of them we should discuss though, as I'm not entirely sure how to proceed.

File: src/building/building.c [hunk 11]
```diff
@@ -713,3 +713,16 @@ void building_load_state(buffer *buf, buffer *sequence, buffer *corrupt_houses,
     extra.incorrect_houses = buffer_read_i32(corrupt_houses);
     extra.unfixable_houses = buffer_read_i32(corrupt_houses);
 }
+
+translation_key building_translation_key(building *b)
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/building/building.c:717]: This is also used in `window/building/depot.c`. 

Building name translations (native and new buildings) will likely be used more in the future. I think this either resides in `building/building.h` or `translation/translation.h`. Thoughts?

File: src/building/building.h [hunk 12]
```diff
@@ -4,6 +4,25 @@
 #include "building/type.h"
 #include "core/buffer.h"
 #include "core/time.h"
+#include "game/resource.h"
+#include "translation/translation.h"
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/building/building.h:8]: Ack.

File: src/building/building.h [hunk 13]
```diff
@@ -260,4 +282,6 @@ void building_save_state(buffer *buf, buffer *highest_id, buffer *highest_id_eve
 
 void building_load_state(buffer *buf, buffer *sequence, buffer *corrupt_houses, int includes_building_size, int save_version);
 
+translation_key building_translation_key(building *b);
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/building/building.h:285]: Ack

File: src/building/building.h [hunk 14]
```diff
@@ -157,6 +176,9 @@ typedef struct building {
         struct {
             short flag_frame;
         } warehouse;
+        struct {
+            order order1;
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/building/building.h:180]: Refactored. This was added when I was unsure how many orders each depot should support.

File: src/figure/figure.h [hunk 15]
```diff
@@ -118,6 +119,7 @@ typedef struct {
         unsigned short visited_building_type_ids[12];
         unsigned char tourist_rank;
     } tourist;
+    order current_order;
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/figure/figure.h:122]: No. Once the cart leaves the depot, it is determined to fulfil its current order. The player can modify order at depot at this point, but the new order will only be picked up by the cart once it finishes/cancels its current order. A save can take place before the cart finishes current order. After loading, the cart should carry on with its current order and then start the new order.

File: src/figuretype/depot.c [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+#include "figure/movement.h"
+#include "figure/route.h"
+#include "game/resource.h"
+#include "map/road_access.h"
+#include "map/road_network.h"
+#include "map/routing.h"
+#include "map/routing_terrain.h"
+#include "figure/image.h"
+
+#define FIGURE_REROUTE_DESTINATION_TICKS 120
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/figuretype/depot.c:22]: I don't remember why I added this macro here, maybe it now looks conflicting due to refactoring. I guess the compiler just overwrites or ignores duplicate macro definition.

Anyway, I removed the definition here, and things still work. Great.

File: src/figuretype/depot.c [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+static const int CART_OFFSET_8_LOADS_FOOD[] = { 0, 40, 48, 56, 0, 0, 64, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+
+static int cartpusher_speed();
+static void update_image(figure *f);
+static void set_cart_graphic(figure *f);
+static int get_storage_road_access(building *b, map_point *point);
+static void dump_goods(figure *f);
+static int is_order_condition_satisfied(building *depot, order *order);
+static int storage_remove_resource(building *b, int resource, int amount);
+static int storage_add_resource(building *b, int resource, int amount);
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/figuretype/depot.c:35]: I've moved the functions around to get rid of these declarations.

File: src/figuretype/depot.c [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+    figure_image_increase_offset(f, 12);
+    f->cart_image_id = 0;
+    int speed_factor = cartpusher_speed();
+    int percentage_speed = 0;
+    int road_network_id = map_road_network_get(f->grid_offset);
+    f->terrain_usage = TERRAIN_USAGE_ROADS;
+    building *b = building_get(f->building_id);
+
+    switch (f->action_state) {
+    case FIGURE_ACTION_150_ATTACK:
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/figuretype/depot.c:48]: It looks indented on my end. This is weird, I don't think it's a tab vs space issue. Will check again after other code formatting changes.

File: src/figuretype/depot.c [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+    if (f->cart_image_id) {
+        f->cart_image_id += dir;
+        figure_image_set_cart_offset(f, dir);
+        if (f->loads_sold_or_carrying >= 8) {
+            f->y_offset_cart -= 40;
+        }
+    }
+}
+
+static int is_order_condition_satisfied(building *depot, order *order)
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/figuretype/depot.c:233]: Yes they both can.

File: src/figuretype/depot.c [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+    } else if (b->type == BUILDING_WAREHOUSE) {
+        if (b->has_road_access == 1) {
+            map_point_store_result(b->x, b->y, point);
+            return 1;
+        }
+        else if (!map_has_road_access_rotation(b->subtype.orientation, b->x, b->y, 3, point)) {
+            return 0;
+        } else {
+            return 1;
+        }
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/figuretype/depot.c:305]: I don't remember why I did it like that, maybe I was referring to some old code that got refactored.

File: src/building/building.h [hunk 21]
```diff
@@ -4,6 +4,25 @@
...
 #include "core/time.h"
+#include "game/resource.h"
+#include "translation/translation.h"
+
+typedef enum order_condition_type {
+    ORDER_CONDITION_NEVER,
+    ORDER_CONDITION_ALWAYS,
+    ORDER_CONDITION_SOURCE_HAS_MORE_THAN,
+    ORDER_CONDITION_DESTINATION_HAS_LESS_THAN
+} order_condition_type;
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/building/building.h:15]: Let's move this discussion to Discord, and the implementation will likely depend on depot mechanism. Currently either way is OK given that order and depot is 1:1 mapping. We may possibly want to change depot mechanism to 1-many or many-many.

File: src/translation/english.c [hunk 22]
```diff
@@ -844,7 +847,23 @@ static translation_string all_strings[] = {
     {TR_CHEAT_FINISHED_MONUMENTS, "Monuments finished"},
     {TR_CHEAT_UPDATED_MONUMENTS, "Monuments updated"},
     {TR_CHEAT_UNLOCKED_ALL_BUILDINGS, "All buildings unlocked"},
-    {TR_CHEAT_INCITED_RIOT, "Incited a riot"}
+    {TR_CHEAT_INCITED_RIOT, "Incited a riot"},
+    {TR_BUILDING_GRANARY, "Granary" },
+    {TR_BUILDING_WAREHOUSE, "Warehouse" },
```
[the remark this answers is not in the document]
Comment (RuoanJi) [reply src/translation/english.c:852]: I'll address this later.

File: src/figure/figure.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (crudelios) [reply src/figure/figure.h:122]: Makes sense!

File: src/building/building.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (crudelios) [reply src/building/building.c:717]: Humm. I think it makes more sense in translation, but this way the building name is also centralised, like its image.

Right now things are done manually for every building, even on the building properties window.

For consistency sake I'd say keep this on translation (or even nowhere - just use magic values when needed, like for all other buildings).

However, your way of thinking is correct. We should centralise this. Maybe add a `lang_group` and `lang_id` to `building/properties` and fill those. This is a work for some other time though.

**Comment by crudelios** [comment]
This is looking good.

The only thing left is to bump the savegame version in `src/game/file_io.c` 👍 

And add a commented `#define SAVE_GAME_LAST_VERSION_WITHOUT_CART_DEPOT <current version>` or something there, just for documentation purpose, please.

**Comment by crudelios** [comment]
Closed due to updated PR on #818.

**Pull request closed without merging**